### DISCUSSION TITLE
Why are the indentation rules so restrictive?

### DISCUSSION THREAD

**Discussion by Rasmus-Bertell** [PHP-CS-Fixer/PHP-CS-Fixer#7810] [Q&A] [answered]
I'm trying to understand why the indentation rules are so limited? I like to follow the linux kernel spec and usually have 8 space indentation for my personal projects. It's frustrating that there's a function for setting the indentation but it only accepts a limited set of options.

**Comment by julienfalque** [comment]
Can you be more specific about the limitation you are talking about? What prevents you from [defining indentation](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/config.rst#configuring-whitespaces) as expected?

**Comment by kubawerlos** [comment]
@julienfalque there is [this](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.49.0/src/WhitespacesFixerConfig.php#L28) limitation to what indent can be.

**Reply by Wirone** [reply]
Does it make sense to limit this? 🤔

**Reply by julienfalque** [reply]
Ah, I didn't look at the right place 😅

I don't see a valid reason to forbid 8 spaces or whatever other value.

**Reply by Rasmus-Bertell** [reply]
There is this [comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/4150#issuecomment-450091633) from a few years back that basically said it is by design, but if it's ok I can create a PR to change it to arbitrary sized spaces.

**Reply by Wirone** [reply]
The last word goes to @keradus but IMHO it does not make sense to limit this. If people want 1 space and everything works with such a setting, why should we prevent it? We can promote standards with the default config anyway 🙂.

**Reply by keradus** [reply]
TL;DR - is there a strong community that recommend another option than currently available?
If yes - we can consider adding that option.
If no - please go with custom rule to fit your personal needs.

----------

long answer

The statement from @SpacePossum is still valid and still have not received an answer.

For this or any other topic, we aim to promote best practices, with awareness that different communities may have different practices.
For that particular rule, we allow the configuration to fit a vast majority of ecosystem standards - 4 spaces and tab, plus also a 2 spaces as maybe not top popular, but still quite present.

I do not want to have in core repo rule/options that is not reflecting best standards, as I do not want to promote them (and in my head sole fact of having them in core repo means we promote them). We also do not have the test coverage for other variants [and yes, having both, 2 and 4 spaces allowed required us to adopt the code to make everything works properly, not limiting to single rule class but to adjustments across multiple code pieces] - so this would be yet another reason to not allow for 7 spaces or even `"   \t "`.

**Reply by Wirone** [reply]
@keradus the answer to @SpacePossum's question is in the discussion - Linux kernel. I know it's not PHP, but does it really matter? Also, I don't understand your reasoning. Indentation is a person/team/project specific thing, it's configurable and totally separated from the rules (I mean: it does not affect the logic, it's only used by indentation-aware rules). I don't see a reason to limit it in any way unless it's causing problems to Fixer's flow (but I believe it's not).

I am in favor of getting rid of this limitation if it does not affect the tool (I will check it later, AFK now).

**Reply by keradus** [reply]
>  it in any way unless it's causing problems to Fixer's flow (but I believe it's not).

we supported 4 spaces initially. added tabs caused problems (not limited to single rule class). then adding 2 spaces also added problems.

I'm against adding anything that I do not consider a good practice unless there is strong community (yes, PHP community).
If we have strong community using 8 spaces we can consider allowing it (with bunch of tests and hoping for no need for other fixes needed - at least mimic `tests/Fixtures/Integration/set/@PSR12_whitespaces.test*.php` but with new indent option). still no value in 7 spaces being allowed.

If individual wants to use some style that is not common, they are fully encouraged to do so, as we allow for custom rules.

**Reply by Rasmus-Bertell** [reply]
> please go with custom rule to fit your personal needs.

I'm not so experienced with this but can a custom rule overwrite the `->setIndent()` setting?

If it can then, while not ideal in my opinion, this can be solved using a custom rule. I would just have to copy the rule to all my projects.

**Reply by keradus** [reply, marked as answer]
very good point.
so, let's conclude on need of community (or lack of it) part, and potentially extend all the tests.
